akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala

```diff
@@ -284,6 +291,51 @@ import akka.stream.stage._
     override def toString: String = "SingleSource"
   }
 
+  final class IterableSource[T](val elements: immutable.Iterable[T]) extends GraphStage[SourceShape[T]] {
+    ReactiveStreamsCompliance.requireNonNullElement(elements)
+    override protected def initialAttributes: Attributes = DefaultAttributes.iterableSource
+    private val out = Outlet[T]("IterableSource.out")
+    override val shape: SourceShape[T] = SourceShape(out)
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+      new GraphStageLogic(shape) with OutHandler {
+        private lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
+        private var currentIterator: Iterator[T] = _
+
+        override def onPull(): Unit =
+          try {
+            if (currentIterator eq null) {
+              currentIterator = elements.iterator
+            }
+            tryPushNextOrComplete()
+          } catch {
+            case NonFatal(ex) =>
+              decider(ex) match {
+                case Supervision.Stop   => failStage(ex)
+                case Supervision.Resume => tryPushNextOrComplete()
+                case Supervision.Restart =>
+                  currentIterator = elements.iterator
+                  tryPushNextOrComplete()
+              }
+          }
+
+        private def tryPushNextOrComplete(): Unit =
+          if (currentIterator.hasNext) {
+            if (isAvailable(out)) {
+              push(out, currentIterator.next())
+            }
+            if (!currentIterator.hasNext) {
```

Comment:
why call hasNext here,is it ok not to to call here?

Reply:
If it just pushed the last element it can immediately complete the stage after that instead of waiting for next pull to make that decision.